File: python/taflow/metrics/average_drawdown.py
"""Whole-history average drawdown episode metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import AverageDrawdown as _Native
from ._input import as_metric_series


class AverageDrawdown:
    """Compute mean trough magnitude across distinct drawdown episodes.

    The independent oracle is the pinned PerformanceAnalytics 2.1.0
    ``AverageDrawdown`` and
    ``findDrawdowns`` with geometric compounding. Wealth begins at a phantom
    value of one. Each contiguous run where percentage drawdown is strictly
    negative is one episode, and the episode contributes the positive magnitude
    of its deepest trough. A return to drawdown zero completes the episode; an
    unrecovered episode at the end is included using its current deepest trough.
    Warm-up ends after one usable return: a non-empty path with no negative
    episode returns ``0.0``. Empty and all-missing states return ``None``. The source convention is pinned to the
    CRAN 2.1.0 tarball with SHA-256
    ``fc801d39382818cd3a7052326b45d078302aef4d290c85dab83498ed4516d58d``.

    Inputs may be decimal simple returns, log returns, positive equity levels,
    or non-cumulative period P&L. The P&L input method requires positive initial
    equity and Rust performs causal capital conversion. The first equity level
    establishes a conversion baseline and does not increment metric length.
    ``nan_policy`` is ``"omit"`` or ``"raise"``; infinities and simple returns
    below -100% are rejected. Mutating lifecycle methods are fluent, native
    bulk work releases the GIL, and Rust owns all conversion and bounded O(1)
    streaming state.
    """

    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(
        self, returns: Any, *, column: str | None = None
    ) -> "AverageDrawdown":
        """Append chronological decimal simple returns and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_log_returns(
        self, log_returns: Any, *, column: str | None = None
    ) -> "AverageDrawdown":
        """Append chronological log returns and return this metric."""
        self._state.from_log_returns(as_metric_series(log_returns, column=column))
        return self

    def from_equity(
        self, equity: Any, *, column: str | None = None
    ) -> "AverageDrawdown":
        """Append chronological positive equity levels and return this metric."""
        self._state.from_equity(as_metric_series(equity, column=column))
        return self

    def from_pnl(
        self,
        pnl: Any,
        initial_capital: float,
        *,
        column: str | None = None,
    ) -> "AverageDrawdown":
        """Append period P&L using required positive initial capital."""
        self._state.from_pnl(
            as_metric_series(pnl, column=column), float(initial_capital)
        )
        return self

    def append(self, value: float) -> "AverageDrawdown":
        """Append one value in the selected domain and return this metric."""
        self._state.append(float(value))
        return self

    def extend(self, values: Any) -> "AverageDrawdown":
        """Append a chronological series in the selected domain and return this metric."""
        self._state.extend(as_metric_series(values))
        return self

    @property
    def value(self) -> float | None:
        """Return the positive mean episode depth, or ``None`` when empty."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "AverageDrawdown":
        """Clear observations, preserve input configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable normalized returns processed by Rust."""
        return len(self._state)


__all__ = ["AverageDrawdown"]

File: crates/taflow-metrics/src/metrics/average_drawdown.rs
use crate::{
    primitives::DrawdownState, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Mean trough magnitude across distinct drawdown episodes.
#[derive(Debug, Clone)]
pub struct AverageDrawdown {
    input: MetricInputState,
    drawdown: DrawdownState,
    completed_depth_sum: f64,
    completed_episode_count: usize,
    current_episode_depth: Option<f64>,
}

impl AverageDrawdown {
    /// Construct an empty state with an explicitly selected semantic input mode.
    pub fn new(input_kind: MetricInputKind, nan_policy: NanPolicy) -> MetricResult<Self> {
        if matches!(
            input_kind,
            MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{input_kind:?}"),
                reason: "average drawdown requires returns, log returns, equity, or period P&L with initial equity",
            });
        }
        Ok(Self {
            input: MetricInputState::new(input_kind, nan_policy)?,
            drawdown: DrawdownState::new(),
            completed_depth_sum: 0.0,
            completed_episode_count: 0,
            current_episode_depth: None,
        })
    }

    /// Append one chronological observation and return the current average.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(simple_return) = self.input.append(value)? {
            self.drawdown.append(simple_return)?;
            let current_drawdown = self
                .drawdown
                .current_drawdown()
                .expect("an appended return always produces a drawdown");
            if current_drawdown < 0.0 {
                self.current_episode_depth = Some(
                    self.current_episode_depth
                        .map_or(current_drawdown, |depth| depth.min(current_drawdown)),
                );
            } else if let Some(depth) = self.current_episode_depth.take() {
                self.completed_depth_sum += depth.abs();
                self.completed_episode_count += 1;
            }
        }
        Ok(self.value())
    }

    /// Append a chronological slice through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        for &value in values {
            self.append(value)?;
        }
        Ok(self.value())
    }

    /// Return the mean positive trough magnitude across observed episodes.
    ///
    /// A current unrecovered episode participates alongside completed episodes.
    /// A non-empty path without a negative drawdown returns zero; an empty path
    /// returns `None`.
    pub fn value(&self) -> Option<f64> {
        if self.input.is_empty() {
            return None;
        }
        if let Some(depth) = self.current_episode_depth {
            return Some(
                (self.completed_depth_sum + depth.abs())
                    / (self.completed_episode_count + 1) as f64,
            );
        }
        if self.completed_episode_count == 0 {
            Some(0.0)
        } else {
            Some(self.completed_depth_sum / self.completed_episode_count as f64)
        }
    }

    /// Return the current result without replaying prior observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving input configuration.
    pub fn reset(&mut self) {
        self.input.reset();
        self.drawdown.reset();
        self.completed_depth_sum = 0.0;
        self.completed_episode_count = 0;
        self.current_episode_depth = None;
    }

    /// Return the number of usable normalized returns processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable normalized returns have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
